inc/functions.php

```diff
@@ -1857,6 +1857,66 @@ function is_moderator($fid=0, $action="", $uid=0)
 	}
 }
 
+/**
+ * Get an array of fids that the forum moderator has access to.
+ * Do not use for administraotrs or global moderators as they moderate any forum and the function will return false.
+ * 
+ * @param int $uid The user ID (0 assumes current user)
+ * @return array|bool an array of the fids the user has moderator access to or bool if called incorrectly.
+ */
+function get_moderated_fids($uid=0)
+{
+	global $mybb, $cache;
+
+	if($uid == 0)
+	{
+		$uid = $mybb->user['uid'];
+	}
+
+	if($uid == 0)
+	{
+		return false;
+	}
+
+	$user_perms = user_permissions($uid);
+
+	if($user_perms['issupermod'] == 1)
+	{
+		return false;
+	}
```

Comment:
Super mods cannot have their moderator permissions revoked for particular forums (even by assigning them as moderator to a forum and unchecking all options).

The intention for returning false here was to strongly discourage use of the function to be used for an super mod. The calling function should be dealing with super mods separately before using `get_moderated_fids` only for standard moderators.

Reply:
Yeah that also makes sense but please note `is_moderator()` and `get_moderator_permissions()` do have different results for super mods.

Reply:
It looks like `get_moderator_permissions()` is only used for non-super moderators in `is_moderator`.

Super mods can moderate any forum with all moderator permissions as long as they can view the forum. Returning false indicates to the caller that they need to handle the super moderator case correctly. For example, the search queries include the unviewable forums condition which covers this problem.

I do think that `get_moderated_fids` should check child forums can be viewed though since it just includes all child forums: https://github.com/mybb/mybb/pull/4091/files#diff-7fe6e55397c77ab9a0f5d57bc4cbe5b9R1900 This isn't a problem now because the the unviewable condition but could be in future when someone else uses this function for a different purpose.

Reply:
Ok I was wrong here since the `moderators` cache is built with consideration of child forums so `get_moderated_fids` just needs to find fids relevant to the current user.

Because the `moderators` [cache build](https://github.com/mybb/mybb/blob/16667c33538343039d2e851913059bd220ba5064/inc/class_datacache.php#L716) doesn't take into account whether the moderator can actually view the forum, the assumption is that it is handled later. So this reaffirms that the callers need to be ensuring the super mod case is checked and that the fid is viewable to the moderator.

Reply:
> Ok I was wrong here since the `moderators` cache is built with consideration of child forums so `get_moderated_fids` just needs to find fids relevant to the current user.

Hmm, I have to say that I'm still doubting if the `moderators` cache is doing well. If it's just fine then we can leave `get_moderated_fids` simple but if it's not doing correctly, for example not having correct moderators listed in child forums, what can we do?

Hint: the cache handler is not doing correctly with this `array_merge()` on https://github.com/mybb/mybb/blob/9feef312dc22abef36dcc1ddb1a703e66d9c2fe1/inc/class_datacache.php#L830

Example:
If `$forum_mods` is 
```
array(
	'users' => array(
		3 => ...
		7 => ...

	)
)
```
and `$this->moderators[$forum['fid']]` is
```
array(
	'users' => array(
		1 => ...
		5 => ...

	)
)
```
the resulting merged array would be
```
array(
	'users' => array(
		1 => ...
		5 => ...

	)
)

Maybe we can as suggested by @euantorano to merge this PR and move discussion to #3996 ?

> Because the `moderators` [cache build](https://github.com/mybb/mybb/blob/16667c33538343039d2e851913059bd220ba5064/inc/class_datacache.php#L716) doesn't take into account whether the moderator can actually view the forum, the assumption is that it is handled later. So this reaffirms that the callers need to be ensuring the super mod case is checked and that the fid is viewable to the moderator.

Yeah that means `is_moderator()` is still needed or the low-level one `get_moderator_permissions()` for getting correct permissions.